### ISSUE TITLE
Task Cancelled exceptions after updating to 2.6.0

### ISSUE THREAD

**Issue by mikkark** [IdentityServer/IdentityServer3#3756]
### Question / Issue
After updating to 2.6.0 we started seeing Task cancelled exceptions in the log. I haven't had much luck trying to track down the root cause for it.

We have several customised services (for claims, events, localisation and views) and the only thing that seems to be affected by this is the event service. There we make an HTTP request to create a user notification about the user logging in. Sometimes this notification creation just does not happen and the Task cancelled exception appears in the logs. However, the exception can appear in the logs also when the notification creation succeeds.

### Relevant parts of the log file

22 Jun 2017 10:57:23.431  2017-06-22 07:57:23,478 INFO Start access token validation request
» 22 Jun 2017 10:57:23.431  2017-06-22 07:57:23,478 INFO Start access token validation
» 22 Jun 2017 10:57:23.433  2017-06-22 07:57:23,572 FATAL identityserver Unhandled exception: sender: System.Threading.Tasks.ContinuationTaskFromTask, (first real exception)System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Web.ThreadContext.AssociateWithCurrentThread(Boolean setImpersonationContext)
   at System.Web.HttpApplication.OnThreadEnterPrivate(Boolean setImpersonationContext)
   at System.Web.HttpApplication.System.Web.Util.ISyncContext.Enter()
   at System.Web.Util.SynchronizationHelper.SafeWrapCallback(Action action)
   at System.Web.Util.SynchronizationHelper.<>c__DisplayClass22_0.<QueueAsynchronous>b__0(Task _)
   at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Web.ThreadContext.AssociateWithCurrentThread(Boolean setImpersonationContext)
   at System.Web.HttpApplication.OnThreadEnterPrivate(Boolean setImpersonationContext)
   at System.Web.HttpApplication.System.Web.Util.ISyncContext.Enter()
   at System.Web.Util.SynchronizationHelper.SafeWrapCallback(Action action)
   at System.Web.Util.SynchronizationHelper.<>c__DisplayClass22_0.<QueueAsynchronous>b__0(Task _)
   at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()<---
» 22 Jun 2017 10:57:23.433  2017-06-22 07:57:23,572 FATAL identityserver Unhandled exception: sender: System.Threading.Tasks.ContinuationTaskFromTask, (first real exception)System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Web.ThreadContext.AssociateWithCurrentThread(Boolean setImpersonationContext)
   at System.Web.HttpApplication.OnThreadEnterPrivate(Boolean setImpersonationContext)
   at System.Web.HttpApplication.System.Web.Util.ISyncContext.Enter()
   at System.Web.Util.SynchronizationHelper.SafeWrapCallback(Action action)
   at System.Web.Util.SynchronizationHelper.<>c__DisplayClass22_0.<QueueAsynchronous>b__0(Task _)
   at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Web.ThreadContext.AssociateWithCurrentThread(Boolean setImpersonationContext)
   at System.Web.HttpApplication.OnThreadEnterPrivate(Boolean setImpersonationContext)
   at System.Web.HttpApplication.System.Web.Util.ISyncContext.Enter()
   at System.Web.Util.SynchronizationHelper.SafeWrapCallback(Action action)
   at System.Web.Util.SynchronizationHelper.<>c__DisplayClass22_0.<QueueAsynchronous>b__0(Task _)
   at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()<---
» 22 Jun 2017 10:57:23.445  2017-06-22 07:57:23,572 INFO Start access token validation request
» 22 Jun 2017 10:57:23.445  2017-06-22 07:57:23,572 INFO Start access token validation
» 22 Jun 2017 10:57:23.459  2017-06-22 07:57:23,590 INFO Start access token validation request
» 22 Jun 2017 10:57:23.459  2017-06-22 07:57:23,590 INFO Start access token validation
» 22 Jun 2017 10:57:23.515  2017-06-22 07:57:23,650 INFO Token validation success

**Comment by mabochster** [comment]
Have you been able to figure this one out? I have been experiencing the same thing, the problem is that I have just started using IdentityServer3 and so I do not have much knowledge about how to go about fixing the issue, I have posted a question on stackoverflow and it doesn't seem to be getting much notice. The link to the question is at [https://stackoverflow.com/questions/44997140/identity-server-3-getstringstringasync-throwing-taskcancelledexception](LINK).
